### PR TITLE
Add DebugValue for invisible store in single_store_elim

### PR DESCRIPTION
**Pull request by jaebaek**
The front-end language compiler would simply emit DebugDeclare for
a variable when it is declared, which is effective through the variable's
scope. Since DebugDeclare only maps an OpVariable to a local variable,
the information can be removed when an optimization pass uses the
loaded value of the variable. DebugValue can be used to specify the
value of a variable. For each value update or phi instruction of a variable,
we can add DebugValue to help debugger inspect the variable at any
point of the program execution.
For example,

float a = 3;
... (complicated cfg) ...
foo(a); // <-- variable inspection: debugger can find DebugValue of `float a` in the nearest dominant

For the code with complicated CFG e.g., for-loop, if-statement, we
need help of ssa-rewrite to analyze the effective value of each variable
in each basic block.

If the value update of the variable happens only once and it dominates
all its uses, local-single-store-elim pass conducts the same value update
with ssa-rewrite and we have to let it add DebugValue for the value assignment.

One main issue is that we have to add DebugValue only when the value
update of a variable is visible to DebugDeclare. For example,

```
{  // scope1
   %stack = OpVariable %ptr_int %int_3
   {  // scope2
       DebugDeclare %foo %stack    <-- local variable "foo" in high-level language source code is declared as OpVariable "%stack"
       // add DebugValue "foo = 3"
       ...
       Store %stack %int_7   <-- foo = 7, add DebugValue "foo = 7"
       ...
       // debugger can inspect the value of "foo"
    }
    Store %stack %int_11   <-- out of "scope2" i.e., scope of "foo". DO NOT add DebugValue "foo = 11"
}
```

However, the initalization of a variable is an exception.
For example, an argument passing of an inlined function must be done out of
the function's scope, but we must add a DebugValue for it.

```
// in HLSL
bar(float arg) { ... }
...
float foo = 3;
bar(foo);

// in SPIR-V
%arg = OpVariable
OpStore %arg %foo   <-- Argument passing. Out of "float arg" scope, but we must add DebugValue for "float arg"
... body of function bar(float arg) ...
```

This PR handles the except case in local-single-store-elim pass. It adds
DebugValue for a store that is considered as an initialization.

The same exception handling code for ssa-rewrite is done by this commit: df4198e50eaa705298167d3c0b3cb01ffd28a5ff.

### CODE REVIEW

File: source/opt/local_single_store_elim_pass.cpp
Comment (kuhar) [inline source/opt/local_single_store_elim_pass.cpp:179]: Can we have more than one decl that we want to add a debug value for?

File: source/opt/local_single_store_elim_pass.cpp
Comment (kuhar) [inline source/opt/local_single_store_elim_pass.cpp:182]: This only happens if `modified` was set to true in the loop, no? If that's the case, I think we don't have to check again here.

File: source/opt/debug_info_manager.cpp
Comment (kuhar) [inline source/opt/debug_info_manager.cpp:529]: What happened with checking if `scope_and_line` is not null?

**Review by kuhar — COMMENTED** [review]
nit: you have a typo in the PR title / the first line of commit message, `s/invisiable/invisible`

File: source/opt/debug_info_manager.cpp
Comment (jaebaek) [reply source/opt/debug_info_manager.cpp:529]: The assert in line [446](https://github.com/KhronosGroup/SPIRV-Tools/pull/4002/files/b1bc0c10a2d83a3852d0b07d803b11c8e6abb525#diff-6e5ffb311b0364a87e9794b09d9efa5d5a431fa651b137dce6b1715c8bc9be53L446) will fail. I want to deny null for `scope_and_line` for now (maybe we want to allow it in the future if I realize there are some corner cases).

I will add the same assert in this function for the readability.

File: source/opt/local_single_store_elim_pass.cpp
Comment (jaebaek) [reply source/opt/local_single_store_elim_pass.cpp:179]: Yes. Some language uses multiple local variables using a single `OpVariable` e.g., a reference variable.

Comment (kuhar) [reply source/opt/local_single_store_elim_pass.cpp:179]: Ack. Thanks.

File: source/opt/debug_info_manager.cpp
Comment (kuhar) [reply source/opt/debug_info_manager.cpp:529]: OK

File: source/opt/local_single_store_elim_pass.cpp
Comment (jaebaek) [reply source/opt/local_single_store_elim_pass.cpp:182]: No. I think it can happens even though `modified` is false.
It sounds weird but some high-level languages can allow declaration of local variables without the initialization.
If there is no proper value assignment at all for the DebugDeclare, this code just removes the DebugDeclare without any DebugValue for it.

Comment (kuhar) [reply source/opt/local_single_store_elim_pass.cpp:182]: OK

**Review by kuhar — APPROVED** [review]
LGTM, but please wait for a second review before submitting.

**Comment by jaebaek** [comment]
Sure. I think @dnovillo can help.

**Review by dnovillo — CHANGES_REQUESTED** [review]
Could you add an example in the description?  I'm having a hard time picturing what the problem is.

**Comment by jaebaek** [comment]
@dnovillo I updated the description. Because of its related background e.g., DebugValue distribution, the description may be hard to understand. I will explain it to you in offline.

**Review by dnovillo — CHANGES_REQUESTED** [review]
Thanks for the added description.  It's now a bit more clear to me.  One other question below.

File: test/opt/local_single_store_elim_test.cpp
Comment (jaebaek) [reply test/opt/local_single_store_elim_test.cpp]: The same with the `OpLine` of the store.

```
OpLine %file 3 7
OpStore %storage %value
...
OpNoLine
DebugDeclare %foo %storage  // local variable "foo" is declared as "%storage"
```

must be

```
OpLine %file 3 7
OpStore %storage %value
OpLine %file 3 7
DebugValue %foo %value // the value of "foo" is "%value" - this DebugValue has the same OpLine with "OpStore %storage %value"
```

Comment (jaebaek) [reply test/opt/local_single_store_elim_test.cpp]: I updated the comment

Comment (dnovillo) [reply test/opt/local_single_store_elim_test.cpp]: Thanks!

**Review by dnovillo — APPROVED** [review]

**Pull request merged**
